src/lib/logical_query_plan/stored_table_node.cpp

```diff
@@ -92,6 +92,48 @@ bool StoredTableNode::is_column_nullable(const ColumnID column_id) const {
   return table->column_is_nullable(column_id);
 }
 
+std::vector<FunctionalDependency> StoredTableNode::functional_dependencies() const {
+  auto fds = std::vector<FunctionalDependency>();
+  const auto& unique_constraints = Hyrise::get().storage_manager.get_table(table_name)->get_soft_unique_constraints();
+  const auto& column_expressions = this->column_expressions();
+
+  for (const auto& constraint : unique_constraints) {
+    // We support FDs for non-nullable columns only
+    if (std::any_of(constraint.columns.cbegin(), constraint.columns.cend(),
+                    [this](const auto column_id) { return this->is_column_nullable(column_id); }))
+      continue;
+
+    auto left = ExpressionUnorderedSet{};
+    auto right = ExpressionUnorderedSet{};
+
+    // Gather column expressions for constraint's column ids
+    for (const auto& expression : column_expressions) {
+      const auto lqp_column = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
+      if (std::any_of(constraint.columns.cbegin(), constraint.columns.cend(),
+                      [&lqp_column](const auto constraint_column_id) {
+                        return constraint_column_id == lqp_column->column_reference.original_column_id();
+                      }))
+        left.insert(lqp_column);
+    }
+
+    Assert(left.size() == constraint.columns.size(),
+           "Failed to collect all column expressions for constraint's column ids");
+
+    // Find column expressions that are functionally dependent
+    std::copy_if(column_expressions.begin(), column_expressions.end(), std::inserter(right, right.begin()),
+                 [&left](std::shared_ptr<AbstractExpression> column_expr) { return !(left.contains(column_expr)); });
+
+    // Create functional dependency
+    if (!right.empty() && std::all_of(right.cbegin(), right.cend(), [this](const auto& right_column_expr) {
+          return !right_column_expr->is_nullable_on_lqp(*this);
```

Comment:
Interesting. I had to think about this one for some time. You seem to be right. 
Since it is not a bidirectional relationship, our current FD definition (Left => Right) is not hurt by null values on the right hand side.
I have to revisit the forwarding logic regarding this. Big thanks for bringing this up!

Reply:
Mmh? This seems to look just like it was before?

Reply:
Thanks! Now it is changed, together with the corresponding test. 
In the course of this, I made `FunctionalDependency` a struct, having a custom `operator==`.